Zmzpwyb 68

221623 opened the pull request.

michal9521 committed:
#time 5h
#comment
- Dodano komunikację z zewnętrznym systemem Yandex do tłumaczenia pytań.

michal9521 committed:
Revert "#time 5h #comment - Dodano komunikację z zewnętrznym systemem Yandex do tłumaczenia pytań."

This reverts commit 496770b

michal9521 committed:
#time 3h
#comment
- Dodano automatyczne tłumaczenie pytań

michal9521 committed:
Revert "#time 5h #comment - Dodano komunikację z zewnętrznym systemem Yandex do tłumaczenia pytań."

This reverts commit 496770b

michal9521 committed:
#time 2h
#comment
- Mała poprawka przy ukrywaniu przycisków automatycznego tłumaczenia w widoku edycji testu
- Dodano język łaciński i esperanto do tłumaczenia
- Wymuszono wybór jednego z dostępnych języków przy tworzeniu testu, w celu uniknięcia błędów w automatycznym tłumaczeniem.

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-questions/test-questions.component.ts:
```diff
@@ -65,6 +68,47 @@ export class TestQuestionsComponent implements OnInit {
     });
   }
 
+  setLanguages(fromLanguage, toLanguage) {
```
W tej metodzie w ogóle niepotrzebne są argumenty, bo wywołujesz ją raz i przekazujesz this.<nazwa_zmiennej>.

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-questions/test-questions.component.ts:
```diff
@@ -65,6 +68,47 @@ export class TestQuestionsComponent implements OnInit {
   }
 
+  setLanguages(fromLanguage, toLanguage) {
+    var fromLang;
```
W TypeScriptcie słowe kluczowym jest `let` lub `const`. `var` jest stosowany w JavaScriptcie.

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-questions/test-questions.component.ts:
```diff
@@ -65,6 +68,47 @@ export class TestQuestionsComponent implements OnInit {
+  setLanguages(fromLanguage, toLanguage) {
+    var fromLang;
+    var toLang;
+    if (fromLanguage === 'polish') {
```
Te całe fragmenty if'owe można wyrzucić do metody.
Zamiast if'ów lepiej wyglądałyby tutaj switch'e.

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-questions/test-questions.component.ts:
```diff
@@ -76,6 +120,50 @@ export class TestQuestionsComponent implements OnInit {
+        if (x.text.length > 0) {
+          targetQuestion.questionPhrase = x.text[0];
+          } else {
+            alert("No translation found");
```
Używasz alerta? Czemu nie naszego serwisu do alertów?

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/tests-create/tests-create.component.html:
```diff
@@ -24,9 +24,17 @@ <h4 class="modal-title">{{title}}</h4>
-      <input type="text" class="form-control input-underline input-lg" placeholder="Language" id="language" name="language" [(ngModel)]="test.language"
-        #language="ngModel" required minlength="3" pattern="^[a-z]+">
-      <div *ngIf="testCreationForm.submitted && language.invalid" class="alert alert-danger">
+        <select class="form-control" name="language" [(ngModel)]="test.language" #language="ngModel" required>
```
Nie można tego wrzucić w tablicę lub mapę w komponencie? W html'u skondensowałoby się to do jednej linijki wtedy i wyglądałoby estetyczniej.

DropsOnly commented:
Dodatkowo:
 - brak plecaholdera na modalu w stylu "Choose target language...",
 - nie powinno być aktualnego języka testu na liście (leci błąd, ale jak go po prostu nie będzie to nie będzie opcji, aby leciał błąd),
 - aktualnie należy przeklikać wszystkie pytania, aby je przetłumaczyć, zalecałbym zmienić to, aby od razu były wszystkie przetłumaczone, gdy userowi się nie spodoba to sobie zmieni (do przedyskutowania ze sztabem rozwojowym),
 - otrzymałem poniższy błąd w wyniku następujących operacji:
    1. Utworzyłem test w języku polskim.
    2. Chciałem go przetłumaczyć na angielski.
    3. Wszystkie pytania kliknąłem, aby się automatycznie przetłumaczyły.
    4. Kliknąłem przycisk "Save".
W bazie brak testu o takiej nazwie i języku angielskim.

![image](https://user-images.githubusercontent.com/19493895/40328850-0f9603fe-5d48-11e8-8362-150dc82de0d4.png)

MaciejZakrzewski committed:
[ZMZPWYB-68] #time 20m #comment Dodano test tworzenia nowej pozycji.

michal9521 committed:
#time 3h
#comment
- Dodano możliwość tłumaczenia wszystkich pytań i opcji na raz
- Dodano placeholdery przy wybieraniu języka
- Zmieniono sposób wyświetlania alertów
-

michal9521 committed:
#time 1h
#comment
- Zmieniono "ify" na "switcha"
- Opcje w drop-downach wczytywane są z obiektu

michal9521 committed:
Merge branch 'ZMZPWYB-68' of https://github.com/miedzinski195659/zwinne-wks into ZMZPWYB-68

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

michal9521 committed:
#time 5M
#comment
- Czyszczenie kodu

DropsOnly commented:
Przycisk moim zdaniem zdecydowanie za duży i nieładnie wygląda w porównaniu do reszty interfejsu.
Dałbym go w nagłówku całego testu.
![image](https://user-images.githubusercontent.com/19493895/40413945-5d0c2d0e-5e77-11e8-9290-0e60a6336fd6.png)

Poniższy fragment ma walidować poprawność selecta? W jaki sposób mogę wpisać cokolwiek do selecta, aby wyskoczył mi komunikat, że za mało znaków wpisałem lub złe znaki?
```
<div class="form-group">
    <select class="form-control" name="language" placeholder="Select target language" [(ngModel)]="language" #lang="ngModel" required>
        <option [ngValue]="undefined" disabled hidden selected> Please select target language </option>
        <option *ngFor="let language of languages; trackBy: trackByFun" [value]="language.id" >{{language.name}}</option>
    </select>
    <div *ngIf="testCreationForm.submitted && lang.invalid" class="alert alert-danger">
        <div *ngIf="lang.errors.required">
          Language can not be empty.
        </div>
        <div *ngIf="lang.errors.minlength">
          Language must have at least 3 characters.
        </div>
        <div *ngIf="lang.errors.pattern">
          Language should contain only lower case letters.
        </div>
    </div>
</div>
```

Poniżej błąd jaki wyskoczył, wiadomość średnia, zmieniłbym na coś typu, że język musi zostać wybrany. I nieładna ta wiadomość, masz odpowiednie klasy css odpowiedzialne za wyświetlenie errorów, np. `border border-danger`, `text-danger`.
![image](https://user-images.githubusercontent.com/19493895/40415430-219186da-5e7b-11e8-9bb2-3c024ea35af5.png)

Analogicznie jak przy tworzeniu testu, języki wszystkie powinny być z małych liter.

Nie jest wyłączona opcja 'Please select target language'. Powinna być wyszarzona. Aktualnie sugeruje, że to jest poprawnie wybrana opcja i może zostać wybrana. Zobaczy jak zostało to zrobione w przypadku wyboru typu pytania. Zrób analogicznie. Będzie ładniej.

W pliku `test-questions.component.html` w jednym miejscu jest litera l luźna. Należy ją usunąć, aby nie pojawiała się pod każdym pytaniem numerycznym. 

Przyciski na ekranie tłumaczenia pytań nie są odpowiednio widoczne. Należy dodać tooltip oraz albo klasę css btn-outline albo jakoś inaczej, aby się wyróżniał w ogóle ten przycisk.

_Z poprzedniego review:_
  - _nie powinno być aktualnego języka testu na liście (leci błąd, ale jak go po prostu nie będzie to nie będzie opcji, aby leciał błąd),_
  - _jeden blok switch możesz wyrzucić do metody i wywołać ją dwukrotnie dla innego parametru, wtedy faktycznie będziesz musiał dać parametr metody, ale będzie mniej kodu, czyli mniej miejsc na błędy_

michal9521 committed:
#time 3h
#comment
- Nie ma możliwości wyboru aktualnego języka testu
- Zmieniono styl przycisków tłumaczenia pytań i opcji
- Zmieniono umiejscowienie i widoczność przycisku tłumaczenia całego testu
- Poprawki w jakości kodu
- Zmieniono styl komunikatów o błędzie podczas wyboru języku do tłumaczenia

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-translation/test-translation.component.ts:
```diff
@@ -15,12 +15,26 @@ export class TestTranslationComponent implements OnInit {
 
   ngOnInit() {
+    this.languages = [
+      {id: "polish", name: "polish"},
```
Skoro id jest takie samo jak name to po co obie zmienne? Nie można tylko listy stringów zrobić?

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/test-translation/test-translation.component.ts:
```diff
@@ -31,6 +45,19 @@ export class TestTranslationComponent implements OnInit {
+  }
+
+  deleteCurrentLanguage(languages) {
+    for (let idx in languages) {
```
Za każdym razem przechodzisz po liście. Nie lepiej wyciągnąć indexOf'em index z tablicy i jeśli jest większy od -1 zesplitować?

DropsOnly commented on wks-recruiter-interface/src/app/layout/tests/tests-create/tests-create.component.ts:
```diff
@@ -17,14 +17,24 @@ export class TestsCreateComponent implements OnInit {
   ngOnInit() {
-
+    this.languages = [
+      {id: "polish", name: "polish"},
```
tak samo jak wyżej, skoro id jest takie samo jako name to nie można zrobić tablicy stringów

DropsOnly commented:
- nie ma czerwonej obwódki wokół selecta w przypadku niepoprawnego wybrania wartości.

Reszta widzę poprawiona. 

Jednak ten przycisk do tłumaczenia wszystkich pytań na całą szerokość i nie ma żadnych marginesów. Nadal nie wygląda to ładnie moim zdaniem. Zobacz może jak w innych widokach przyciski są dodawane w nagłówkach, np. w dodawaniu użytkownika czy testu. A jeśli nie to chociaż marginesy jakieś pododawaj, aby jakoś ten przycisk się ładnie prezentował.

Odnośnie przycisku do tłumaczenia pojedynczego pytania. Moim zdaniem teraz wygląda jak wyłączony i nieklikalny. Może spróbuj wrócić do poprzedniego ustawienia i samemu poprzez klasę css dodać ramkę?

Kolejne komentarze do PR.

michal9521 committed:
#time 1h
#comment
- Dodano czerwone ramki w przypadku źle wypełnionych pól i źle wybranych elementów list
- Poprawiono czytelność przycisków w widoku tłumaczenia
- Poprawiono jakość kodu

DropsOnly commented:
Jak widzisz na konfie, grupie nie podoba się za bardzo ten widok. Do poprawy. Skonsultuj się może z kimś jak by on to widział, np. @lmiedzinski. 

Z tego co widzę zmieniałeś tworzenie testu. Nie znika napis `Language must be chosen.` po tym jak wybiorę język poprawny z listy. (widok tworzenia testu bez pytań)

Reszta ok.

michal9521 committed:
#time 30m
#comment
- Przy tworzeniu testu naprawiono komunikat przy wybieraniu języka
- Zmieniono umiejscowienie przycisku tłumaczenia wszystkich pytań

michal9521 committed:
#time 5m
#comment
- Poprawiona responsywność

DropsOnly reviewed (approved):
Wszystko wydaje się teraz okej.
